File: mem0/vector_stores/azure_ai_search.py
import json
import logging
from typing import List, Optional

from pydantic import BaseModel

from mem0.vector_stores.base import VectorStoreBase

try:
    from azure.core.credentials import AzureKeyCredential
    from azure.core.exceptions import ResourceNotFoundError
    from azure.search.documents import SearchClient
    from azure.search.documents.indexes import SearchIndexClient
    from azure.search.documents.indexes.models import (
        HnswAlgorithmConfiguration,
        ScalarQuantizationCompression,
        SearchField,
        SearchFieldDataType,
        SearchIndex,
        SimpleField,
        VectorSearch,
        VectorSearchProfile,
    )
    from azure.search.documents.models import VectorizedQuery
except ImportError:
    raise ImportError(
        "The 'azure-search-documents' library is required. Please install it using 'pip install azure-search-documents==11.5.1'."
    )

logger = logging.getLogger(__name__)


class OutputData(BaseModel):
    id: Optional[str]
    score: Optional[float]
    payload: Optional[dict]


class AzureAISearch(VectorStoreBase):
    def __init__(self, service_name, collection_name, api_key, embedding_model_dims, use_compression):
        """Initialize the Azure Cognitive Search vector store.

        Args:
            service_name (str): Azure Cognitive Search service name.
            collection_name (str): Index name.
            api_key (str): API key for the Azure Cognitive Search service.
            embedding_model_dims (int): Dimension of the embedding vector.
            use_compression (bool): Use scalar quantization vector compression
        """
        self.index_name = collection_name
        self.collection_name = collection_name
        self.embedding_model_dims = embedding_model_dims
        self.use_compression = use_compression
        self.search_client = SearchClient(
            endpoint=f"https://{service_name}.search.windows.net",
            index_name=self.index_name,
            credential=AzureKeyCredential(api_key),
        )
        self.index_client = SearchIndexClient(
            endpoint=f"https://{service_name}.search.windows.net", credential=AzureKeyCredential(api_key)
        )
        self.create_col()  # create the collection / index

    def create_col(self):
        """Create a new index in Azure Cognitive Search."""
        vector_dimensions = self.embedding_model_dims  # Set this to the number of dimensions in your vector

        if self.use_compression:
            vector_type = "Collection(Edm.Half)"
            compression_name = "myCompression"
            compression_configurations = [ScalarQuantizationCompression(compression_name=compression_name)]
        else:
            vector_type = "Collection(Edm.Single)"
            compression_name = None
            compression_configurations = []

        fields = [
            SimpleField(name="id", type=SearchFieldDataType.String, key=True),
            SimpleField(name="user_id", type=SearchFieldDataType.String, filterable=True),
            SimpleField(name="run_id", type=SearchFieldDataType.String, filterable=True),
            SimpleField(name="agent_id", type=SearchFieldDataType.String, filterable=True),
            SearchField(
                name="vector",
                type=vector_type,
                searchable=True,
                vector_search_dimensions=vector_dimensions,
                vector_search_profile_name="my-vector-config",
            ),
            SimpleField(name="payload", type=SearchFieldDataType.String, searchable=True),
        ]

        vector_search = VectorSearch(
            profiles=[
                VectorSearchProfile(name="my-vector-config", algorithm_configuration_name="my-algorithms-config")
            ],
            algorithms=[HnswAlgorithmConfiguration(name="my-algorithms-config")],
            compressions=compression_configurations,
        )
        index = SearchIndex(name=self.index_name, fields=fields, vector_search=vector_search)
        self.index_client.create_or_update_index(index)

    def _generate_document(self, vector, payload, id):
        document = {"id": id, "vector": vector, "payload": json.dumps(payload)}
        # Extract additional fields if they exist
        for field in ["user_id", "run_id", "agent_id"]:
            if field in payload:
                document[field] = payload[field]
        return document

    def insert(self, vectors, payloads=None, ids=None):
        """Insert vectors into the index.

        Args:
            vectors (List[List[float]]): List of vectors to insert.
            payloads (List[Dict], optional): List of payloads corresponding to vectors.
            ids (List[str], optional): List of IDs corresponding to vectors.
        """
        logger.info(f"Inserting {len(vectors)} vectors into index {self.index_name}")

        documents = [
            self._generate_document(vector, payload, id) for id, vector, payload in zip(ids, vectors, payloads)
        ]
        self.search_client.upload_documents(documents)

    def _build_filter_expression(self, filters):
        filter_conditions = []
        for key, value in filters.items():
            # If the value is a string, add quotes
            if isinstance(value, str):
                condition = f"{key} eq '{value}'"
            else:
                condition = f"{key} eq {value}"
            filter_conditions.append(condition)
        # Use 'and' to join multiple conditions
        filter_expression = " and ".join(filter_conditions)
        return filter_expression

    def search(self, query, limit=5, filters=None):
        """Search for similar vectors.

        Args:
            query (List[float]): Query vectors.
            limit (int, optional): Number of results to return. Defaults to 5.
            filters (Dict, optional): Filters to apply to the search. Defaults to None.

        Returns:
            list: Search results.
        """
        # Build filter expression
        filter_expression = None
        if filters:
            filter_expression = self._build_filter_expression(filters)

        vector_query = VectorizedQuery(vector=query, k_nearest_neighbors=limit, fields="vector")
        search_results = self.search_client.search(vector_queries=[vector_query], filter=filter_expression, top=limit)

        results = []
        for result in search_results:
            payload = json.loads(result["payload"])
            results.append(OutputData(id=result["id"], score=result["@search.score"], payload=payload))
        return results

    def delete(self, vector_id):
        """Delete a vector by ID.

        Args:
            vector_id (str): ID of the vector to delete.
        """
        self.search_client.delete_documents(documents=[{"id": vector_id}])
        logger.info(f"Deleted document with ID '{vector_id}' from index '{self.index_name}'.")

    def update(self, vector_id, vector=None, payload=None):
        """Update a vector and its payload.

        Args:
            vector_id (str): ID of the vector to update.
            vector (List[float], optional): Updated vector.
            payload (Dict, optional): Updated payload.
        """
        document = {"id": vector_id}
        if vector:
            document["vector"] = vector
        if payload:
            json_payload = json.dumps(payload)
            document["payload"] = json_payload
            for field in ["user_id", "run_id", "agent_id"]:
                document[field] = payload.get(field)
        self.search_client.merge_or_upload_documents(documents=[document])

    def get(self, vector_id) -> OutputData:
        """Retrieve a vector by ID.

        Args:
            vector_id (str): ID of the vector to retrieve.

        Returns:
            OutputData: Retrieved vector.
        """
        try:
            result = self.search_client.get_document(key=vector_id)
        except ResourceNotFoundError:
            return None
        return OutputData(id=result["id"], score=None, payload=json.loads(result["payload"]))

    def list_cols(self) -> List[str]:
        """List all collections (indexes).

        Returns:
            List[str]: List of index names.
        """
        indexes = self.index_client.list_indexes()
        return [index.name for index in indexes]

    def delete_col(self):
        """Delete the index."""
        self.index_client.delete_index(self.index_name)

    def col_info(self):
        """Get information about the index.

        Returns:
            Dict[str, Any]: Index information.
        """
        index = self.index_client.get_index(self.index_name)
        return {"name": index.name, "fields": index.fields}

    def list(self, filters=None, limit=100):
        """List all vectors in the index.

        Args:
            filters (Dict, optional): Filters to apply to the list.
            limit (int, optional): Number of vectors to return. Defaults to 100.

        Returns:
            List[OutputData]: List of vectors.
        """
        filter_expression = None
        if filters:
            filter_expression = self._build_filter_expression(filters)

        search_results = self.search_client.search(search_text="*", filter=filter_expression, top=limit)
        results = []
        for result in search_results:
            payload = json.loads(result["payload"])
            results.append(OutputData(id=result["id"], score=result["@search.score"], payload=payload))

        return [results]

    def __del__(self):
        """Close the search client when the object is deleted."""
        self.search_client.close()
        self.index_client.close()

File: mem0-ts/src/oss/src/vector_stores/azure_ai_search.ts
import {
  SearchClient,
  SearchIndexClient,
  AzureKeyCredential,
  SearchIndex,
  SearchField,
  SearchFieldDataType,
  SimpleField,
  VectorSearch,
  VectorSearchProfile,
  HnswAlgorithmConfiguration,
  ScalarQuantizationCompression,
  BinaryQuantizationCompression,
  VectorizedQuery,
} from "@azure/search-documents";
import { DefaultAzureCredential } from "@azure/identity";
import { VectorStore } from "./base";
import { SearchFilters, VectorStoreConfig, VectorStoreResult } from "../types";

/**
 * Configuration interface for Azure AI Search vector store
 */
interface AzureAISearchConfig extends VectorStoreConfig {
  /**
   * Azure AI Search service name (e.g., "my-search-service")
   */
  serviceName: string;

  /**
   * Index/collection name to use
   */
  collectionName: string;

  /**
   * API key for authentication (if not provided, uses DefaultAzureCredential)
   */
  apiKey?: string;

  /**
   * Vector embedding dimensions
   */
  embeddingModelDims: number;

  /**
   * Compression type: 'none', 'scalar', or 'binary'
   * @default 'none'
   */
  compressionType?: "none" | "scalar" | "binary";

  /**
   * Use half precision (float16) instead of full precision (float32)
   * @default false
   */
  useFloat16?: boolean;

  /**
   * Enable hybrid search (combines vector + text search)
   * @default false
   */
  hybridSearch?: boolean;

  /**
   * Vector filter mode: 'preFilter' or 'postFilter'
   * @default 'preFilter'
   */
  vectorFilterMode?: string;
}

/**
 * Azure AI Search vector store implementation
 * Supports vector search with hybrid search, compression, and filtering
 */
export class AzureAISearch implements VectorStore {
  private searchClient: SearchClient<any>;
  private indexClient: SearchIndexClient;
  private readonly serviceName: string;
  private readonly indexName: string;
  private readonly embeddingModelDims: number;
  private readonly compressionType: "none" | "scalar" | "binary";
  private readonly useFloat16: boolean;
  private readonly hybridSearch: boolean;
  private readonly vectorFilterMode: string;
  private readonly apiKey: string | undefined;

  constructor(config: AzureAISearchConfig) {
    this.serviceName = config.serviceName;
    this.indexName = config.collectionName;
    this.embeddingModelDims = config.embeddingModelDims;
    this.compressionType = config.compressionType || "none";
    this.useFloat16 = config.useFloat16 || false;
    this.hybridSearch = config.hybridSearch || false;
    this.vectorFilterMode = config.vectorFilterMode || "preFilter";
    this.apiKey = config.apiKey;

    const serviceEndpoint = `https://${this.serviceName}.search.windows.net`;

    // Determine authentication: API key or DefaultAzureCredential
    const credential =
      this.apiKey && this.apiKey !== "" && this.apiKey !== "your-api-key"
        ? new AzureKeyCredential(this.apiKey)
        : new DefaultAzureCredential();

    // Initialize clients
    this.searchClient = new SearchClient(
      serviceEndpoint,
      this.indexName,
      credential,
    );

    this.indexClient = new SearchIndexClient(serviceEndpoint, credential);

    // Initialize the index
    this.initialize().catch(console.error);
  }

  /**
   * Initialize the Azure AI Search index if it doesn't exist
   */
  async initialize(): Promise<void> {
    try {
      const collections = await this.listCols();
      if (!collections.includes(this.indexName)) {
        await this.createCol();
      }
    } catch (error) {
      console.error("Error initializing Azure AI Search:", error);
      throw error;
    }
  }

  /**
   * Create a new index in Azure AI Search
   */
  private async createCol(): Promise<void> {
    // Determine vector type based on use_float16 setting
    const vectorType = this.useFloat16
      ? "Collection(Edm.Half)"
      : "Collection(Edm.Single)";

    // Configure compression settings
    const compressionConfigurations: Array<
      ScalarQuantizationCompression | BinaryQuantizationCompression
    > = [];
    let compressionName: string | undefined = undefined;

    if (this.compressionType === "scalar") {
      compressionName = "myCompression";
      compressionConfigurations.push({
        kind: "scalarQuantization",
        compressionName: compressionName,
      } as ScalarQuantizationCompression);
    } else if (this.compressionType === "binary") {
      compressionName = "myCompression";
      compressionConfigurations.push({
        kind: "binaryQuantization",
        compressionName: compressionName,
      } as BinaryQuantizationCompression);
    }

    // Define index fields
    const fields: SearchField[] = [
      {
        name: "id",
        type: "Edm.String",
        key: true,
      } as SimpleField,
      {
        name: "user_id",
        type: "Edm.String",
        filterable: true,
      } as SimpleField,
      {
        name: "run_id",
        type: "Edm.String",
        filterable: true,
      } as SimpleField,
      {
        name: "agent_id",
        type: "Edm.String",
        filterable: true,
      } as SimpleField,
      {
        name: "vector",
        type: vectorType as SearchFieldDataType,
        searchable: true,
        vectorSearchDimensions: this.embeddingModelDims,
        vectorSearchProfileName: "my-vector-config",
      } as SearchField,
      {
        name: "payload",
        type: "Edm.String",
        searchable: true,
      } as SearchField,
    ];

    // Configure vector search
    const vectorSearch: VectorSearch = {
      profiles: [
        {
          name: "my-vector-config",
          algorithmConfigurationName: "my-algorithms-config",
          compressionName:
            this.compressionType !== "none" ? compressionName : undefined,
        } as VectorSearchProfile,
      ],
      algorithms: [
        {
          kind: "hnsw",
          name: "my-algorithms-config",
        } as HnswAlgorithmConfiguration,
      ],
      compressions: compressionConfigurations,
    };

    // Create index
    const index: SearchIndex = {
      name: this.indexName,
      fields,
      vectorSearch,
    };

    await this.indexClient.createOrUpdateIndex(index);
  }

  /**
   * Generate a document for insertion
   */
  private generateDocument(
    vector: number[],
    payload: Record<string, any>,
    id: string,
  ): Record<string, any> {
    const document: Record<string, any> = {
      id,
      vector,
      payload: JSON.stringify(payload),
    };

    // Extract additional fields if they exist
    for (const field of ["user_id", "run_id", "agent_id"]) {
      if (field in payload) {
        document[field] = payload[field];
      }
    }

    return document;
  }

  /**
   * Insert vectors into the index
   */
  async insert(
    vectors: number[][],
    ids: string[],
    payloads: Record<string, any>[],
  ): Promise<void> {
    console.log(
      `Inserting ${vectors.length} vectors into index ${this.indexName}`,
    );

    const documents = vectors.map((vector, idx) =>
      this.generateDocument(vector, payloads[idx] || {}, ids[idx]),
    );

    const response = await this.searchClient.uploadDocuments(documents);

    // Check for errors
    for (const result of response.results) {
      if (!result.succeeded) {
        throw new Error(
          `Insert failed for document ${result.key}: ${result.errorMessage}`,
        );
      }
    }
  }

  /**
   * Sanitize filter keys to remove non-alphanumeric characters
   */
  private sanitizeKey(key: string): string {
    return key.replace(/[^\w]/g, "");
  }

  /**
   * Build OData filter expression from SearchFilters
   */
  private buildFilterExpression(filters: SearchFilters): string {
    const filterConditions: string[] = [];

    for (const [key, value] of Object.entries(filters)) {
      const safeKey = this.sanitizeKey(key);

      if (typeof value === "string") {
        // Escape single quotes in string values
        const safeValue = value.replace(/'/g, "''");
        filterConditions.push(`${safeKey} eq '${safeValue}'`);
      } else {
        filterConditions.push(`${safeKey} eq ${value}`);
      }
    }

    return filterConditions.join(" and ");
  }

  /**
   * Extract JSON from payload string
   * Handles cases where payload might have extra text
   */
  private extractJson(payload: string): string {
    try {
      // Try to parse as-is first
      JSON.parse(payload);
      return payload;
    } catch {
      // If that fails, try to extract JSON object
      const match = payload.match(/\{.*\}/s);
      return match ? match[0] : payload;
    }
  }

  /**
   * Search for similar vectors
   */
  async search(
    query: number[],
    limit: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[]> {
    const filterExpression = filters
      ? this.buildFilterExpression(filters)
      : undefined;

    const vectorQuery: VectorizedQuery<any> = {
      kind: "vector",
      vector: query,
      kNearestNeighborsCount: limit,
      fields: ["vector"],
    };

    let searchResults;

    if (this.hybridSearch) {
      // Hybrid search: combines vector + text search
      searchResults = await this.searchClient.search("*", {
        vectorSearchOptions: {
          queries: [vectorQuery],
          filterMode: this.vectorFilterMode as any,
        },
        filter: filterExpression,
        top: limit,
        searchFields: ["payload"],
      });
    } else {
      // Pure vector search
      searchResults = await this.searchClient.search("*", {
        vectorSearchOptions: {
          queries: [vectorQuery],
          filterMode: this.vectorFilterMode as any,
        },
        filter: filterExpression,
        top: limit,
      });
    }

    const results: VectorStoreResult[] = [];

    for await (const result of searchResults.results) {
      const payloadStr = result.document.payload as string;
      const payload = JSON.parse(this.extractJson(payloadStr));

      results.push({
        id: result.document.id as string,
        score: result.score,
        payload,
      });
    }

    return results;
  }

  /**
   * Delete a vector by ID
   */
  async delete(vectorId: string): Promise<void> {
    const response = await this.searchClient.deleteDocuments([
      { id: vectorId },
    ]);

    for (const result of response.results) {
      if (!result.succeeded) {
        throw new Error(
          `Delete failed for document ${vectorId}: ${result.errorMessage}`,
        );
      }
    }

    console.log(
      `Deleted document with ID '${vectorId}' from index '${this.indexName}'.`,
    );
  }

  /**
   * Update a vector and its payload
   */
  async update(
    vectorId: string,
    vector: number[],
    payload: Record<string, any>,
  ): Promise<void> {
    const document: Record<string, any> = { id: vectorId };

    if (vector) {
      document.vector = vector;
    }

    if (payload) {
      document.payload = JSON.stringify(payload);

      // Extract additional fields
      for (const field of ["user_id", "run_id", "agent_id"]) {
        if (field in payload) {
          document[field] = payload[field];
        }
      }
    }

    const response = await this.searchClient.mergeOrUploadDocuments([document]);

    for (const result of response.results) {
      if (!result.succeeded) {
        throw new Error(
          `Update failed for document ${vectorId}: ${result.errorMessage}`,
        );
      }
    }
  }

  /**
   * Retrieve a vector by ID
   */
  async get(vectorId: string): Promise<VectorStoreResult | null> {
    try {
      const result = await this.searchClient.getDocument(vectorId);
      const payloadStr = result.payload as string;
      const payload = JSON.parse(this.extractJson(payloadStr));

      return {
        id: result.id as string,
        payload,
      };
    } catch (error: any) {
      // Return null if document not found
      if (error?.statusCode === 404) {
        return null;
      }
      throw error;
    }
  }

  /**
   * List all collections (indexes)
   */
  private async listCols(): Promise<string[]> {
    const names: string[] = [];

    for await (const index of this.indexClient.listIndexes()) {
      names.push(index.name);
    }

    return names;
  }

  /**
   * Delete the index
   */
  async deleteCol(): Promise<void> {
    await this.indexClient.deleteIndex(this.indexName);
  }

  /**
   * Get information about the index
   */
  private async colInfo(): Promise<{ name: string; fields: SearchField[] }> {
    const index = await this.indexClient.getIndex(this.indexName);
    return {
      name: index.name,
      fields: index.fields,
    };
  }

  /**
   * List all vectors in the index
   */
  async list(
    filters?: SearchFilters,
    limit: number = 100,
  ): Promise<[VectorStoreResult[], number]> {
    const filterExpression = filters
      ? this.buildFilterExpression(filters)
      : undefined;

    const searchResults = await this.searchClient.search("*", {
      filter: filterExpression,
      top: limit,
    });

    const results: VectorStoreResult[] = [];

    for await (const result of searchResults.results) {
      const payloadStr = result.document.payload as string;
      const payload = JSON.parse(this.extractJson(payloadStr));

      results.push({
        id: result.document.id as string,
        score: result.score,
        payload,
      });
    }

    return [results, results.length];
  }

  /**
   * Generate a random user ID
   */
  private generateUUID(): string {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(
      /[xy]/g,
      function (c) {
        const r = (Math.random() * 16) | 0;
        const v = c === "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      },
    );
  }

  /**
   * Get user ID from memory_migrations collection
   * Required by VectorStore interface
   */
  async getUserId(): Promise<string> {
    try {
      // Check if memory_migrations index exists
      const collections = await this.listCols();
      const migrationIndexExists = collections.includes("memory_migrations");

      if (!migrationIndexExists) {
        // Create memory_migrations index
        const migrationIndex: SearchIndex = {
          name: "memory_migrations",
          fields: [
            {
              name: "id",
              type: "Edm.String",
              key: true,
            } as SimpleField,
            {
              name: "user_id",
              type: "Edm.String",
              searchable: false,
              filterable: true,
            } as SimpleField,
          ],
        };
        await this.indexClient.createOrUpdateIndex(migrationIndex);
      }

      // Try to get existing user_id
      const searchResults = await this.searchClient.search("*", {
        top: 1,
      });

      for await (const result of searchResults.results) {
        const userId = result.document.user_id as string;
        if (userId) {
          return userId;
        }
      }

      // Generate a random user_id if none exists
      const randomUserId =
        Math.random().toString(36).substring(2, 15) +
        Math.random().toString(36).substring(2, 15);

      await this.searchClient.uploadDocuments([
        {
          id: this.generateUUID(),
          user_id: randomUserId,
        },
      ]);

      return randomUserId;
    } catch (error) {
      console.error("Error getting user ID:", error);
      throw error;
    }
  }

  /**
   * Set user ID in memory_migrations collection
   * Required by VectorStore interface
   */
  async setUserId(userId: string): Promise<void> {
    try {
      // Get existing point ID or generate new one
      const searchResults = await this.searchClient.search("*", {
        top: 1,
      });

      let pointId = this.generateUUID();

      for await (const result of searchResults.results) {
        pointId = result.document.id as string;
        break;
      }

      await this.searchClient.mergeOrUploadDocuments([
        {
          id: pointId,
          user_id: userId,
        },
      ]);
    } catch (error) {
      console.error("Error setting user ID:", error);
      throw error;
    }
  }

  /**
   * Reset the index by deleting and recreating it
   */
  async reset(): Promise<void> {
    console.log(`Resetting index ${this.indexName}...`);

    try {
      // Delete the index
      await this.deleteCol();

      // Recreate the index
      await this.createCol();
    } catch (error) {
      console.error(`Error resetting index ${this.indexName}:`, error);
      throw error;
    }
  }
}
